File: BackEnd/PHP/laravelshop_eloquentorm/tests/Feature/Api/CommunityInquirySupportApiTest.php
<?php

namespace Tests\Feature\Api;

use App\Models\Board;
use App\Models\Category;
use App\Models\Product;
use App\Models\User;
use App\Modules\Auth\Services\JwtService;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Support\Facades\Hash;
use Tests\TestCase;

class CommunityInquirySupportApiTest extends TestCase
{
    use RefreshDatabase;

    public function test_user_can_manage_community_posts_likes_and_comments(): void
    {
        $user = $this->createUser('USER');
        $board = Board::query()->create([
            'name' => '자유게시판',
            'slug' => 'free-board',
            'description' => 'PBShop 커뮤니티',
        ]);

        $boardsResponse = $this->getJson('/api/v1/boards');
        $boardsResponse->assertOk();
        $boardsResponse->assertJsonPath('data.0.slug', 'free-board');

        $createPost = $this->actingAsApiUser($user)->postJson('/api/v1/boards/'.$board->id.'/posts', [
            'title' => '첫 글입니다',
            'content' => '커뮤니티 테스트 본문',
        ]);
        $createPost->assertCreated();
        $postId = $createPost->json('data.id');

        $postsResponse = $this->getJson('/api/v1/boards/'.$board->id.'/posts');
        $postsResponse->assertOk();
        $postsResponse->assertJsonPath('data.items.0.title', '첫 글입니다');

        $showPost = $this->getJson('/api/v1/posts/'.$postId);
        $showPost->assertOk();
        $this->assertEquals(1, $showPost->json('data.viewCount'));

        $likeResponse = $this->actingAsApiUser($user)->postJson('/api/v1/posts/'.$postId.'/like');
        $likeResponse->assertOk();
        $likeResponse->assertJsonPath('data.liked', true);

        $createComment = $this->actingAsApiUser($user)->postJson('/api/v1/posts/'.$postId.'/comments', [
            'content' => '댓글 테스트',
        ]);
        $createComment->assertCreated();
        $commentId = $createComment->json('data.id');

        $listComments = $this->getJson('/api/v1/posts/'.$postId.'/comments');
        $listComments->assertOk();
        $listComments->assertJsonPath('data.0.content', '댓글 테스트');

        $updatePost = $this->actingAsApiUser($user)->patchJson('/api/v1/posts/'.$postId, [
            'title' => '수정된 글',
        ]);
        $updatePost->assertOk();
        $updatePost->assertJsonPath('data.title', '수정된 글');

        $deleteComment = $this->actingAsApiUser($user)->deleteJson('/api/v1/comments/'.$commentId);
        $deleteComment->assertOk();
        $deleteComment->assertJsonPath('data.message', '댓글이 삭제되었습니다.');

        $deletePost = $this->actingAsApiUser($user)->deleteJson('/api/v1/posts/'.$postId);
        $deletePost->assertOk();
        $deletePost->assertJsonPath('data.message', '게시글이 삭제되었습니다.');
    }

    public function test_user_can_create_and_admin_can_answer_product_inquiry(): void
    {
        $user = $this->createUser('USER');
        $outsider = $this->createUser('USER');
        $admin = $this->createUser('ADMIN');
        $category = Category::query()->create([
            'name' => '모니터',
            'slug' => 'monitor',
            'depth' => 0,
            'sort_order' => 0,
            'is_visible' => true,
        ]);
        $product = Product::query()->create([
            'category_id' => $category->id,
            'name' => 'PB Monitor',
            'slug' => 'pb-monitor',
            'status' => 'ACTIVE',
        ]);

        $createInquiry = $this->actingAsApiUser($user)->postJson('/api/v1/products/'.$product->id.'/inquiries', [
            'title' => '재입고 문의',
            'content' => '언제 재입고되나요?',
            'isSecret' => true,
        ]);
        $createInquiry->assertCreated();
        $inquiryId = $createInquiry->json('data.id');

        $listMine = $this->actingAsApiUser($user)->getJson('/api/v1/inquiries/me');
        $listMine->assertOk();
        $listMine->assertJsonPath('data.0.title', '재입고 문의');

        $publicList = $this->actingAsApiUser($outsider)->getJson('/api/v1/products/'.$product->id.'/inquiries');
        $publicList->assertOk();
        $publicList->assertJsonPath('data.0.content', '비밀 문의입니다.');

        $answerInquiry = $this->actingAsApiUser($admin)->postJson('/api/v1/inquiries/'.$inquiryId.'/answer', [
            'answer' => '다음 주 입고 예정입니다.',
        ]);
        $answerInquiry->assertOk();
        $answerInquiry->assertJsonPath('data.answer', '다음 주 입고 예정입니다.');

        $deleteInquiry = $this->actingAsApiUser($user)->deleteJson('/api/v1/inquiries/'.$inquiryId);
        $deleteInquiry->assertOk();
        $deleteInquiry->assertJsonPath('data.message', '문의글이 삭제되었습니다.');
    }

    public function test_user_and_admin_can_handle_support_ticket_flow(): void
    {
        $user = $this->createUser('USER');
        $admin = $this->createUser('ADMIN');

        $createTicket = $this->actingAsApiUser($user)->postJson('/api/v1/support/tickets', [
            'category' => 'delivery',
            'title' => '배송 문의',
            'content' => '배송이 지연되고 있습니다.',
        ]);
        $createTicket->assertCreated();
        $ticketId = $createTicket->json('data.id');
        $createTicket->assertJsonPath('data.status', 'OPEN');

        $listTickets = $this->actingAsApiUser($user)->getJson('/api/v1/support/tickets');
        $listTickets->assertOk();
        $listTickets->assertJsonPath('data.0.title', '배송 문의');

        $adminReply = $this->actingAsApiUser($admin)->postJson('/api/v1/support/tickets/'.$ticketId.'/reply', [
            'content' => '오늘 중으로 확인 후 안내드리겠습니다.',
        ]);
        $adminReply->assertOk();
        $adminReply->assertJsonPath('data.status', 'ANSWERED');
        $adminReply->assertJsonPath('data.replies.0.content', '오늘 중으로 확인 후 안내드리겠습니다.');

        $adminList = $this->actingAsApiUser($admin)->getJson('/api/v1/admin/support/tickets');
        $adminList->assertOk();
        $adminList->assertJsonPath('data.0.ticketNumber', $createTicket->json('data.ticketNumber'));

        $updateStatus = $this->actingAsApiUser($admin)->patchJson('/api/v1/admin/support/tickets/'.$ticketId.'/status', [
            'status' => 'RESOLVED',
        ]);
        $updateStatus->assertOk();
        $updateStatus->assertJsonPath('data.status', 'RESOLVED');

        $showTicket = $this->actingAsApiUser($user)->getJson('/api/v1/support/tickets/'.$ticketId);
        $showTicket->assertOk();
        $showTicket->assertJsonPath('data.status', 'RESOLVED');
    }

    private function createUser(string $role): User
    {
        return User::query()->create([
            'email' => uniqid('cis-', true).'@example.com',
            'password' => Hash::make('Password123!'),
            'name' => 'Community Inquiry Support User',
            'nickname' => 'cis-user',
            'role' => $role,
            'status' => 'ACTIVE',
            'phone' => '01012345678',
            'email_verified_at' => now(),
        ]);
    }

    private function actingAsApiUser(User $user): self
    {
        $token = app(JwtService::class)->createAccessToken($user);

        return $this->withHeader('Authorization', 'Bearer '.$token);
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/test/java/com/pbshop/java/spring/maven/jpa/postgresql/community/CommunityInquirySupportApiTest.java
package com.pbshop.java.spring.maven.jpa.postgresql.community;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.math.BigDecimal;
import java.util.UUID;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.transaction.Transactional;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MvcResult;

import com.pbshop.java.spring.maven.jpa.postgresql.category.Category;
import com.pbshop.java.spring.maven.jpa.postgresql.category.CategoryRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.inquiry.ProductInquiryRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.product.PriceEntry;
import com.pbshop.java.spring.maven.jpa.postgresql.product.PriceEntryRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Product;
import com.pbshop.java.spring.maven.jpa.postgresql.product.ProductRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.product.Seller;
import com.pbshop.java.spring.maven.jpa.postgresql.product.SellerRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.support.ApiIntegrationSupport;
import com.pbshop.java.spring.maven.jpa.postgresql.support.SupportTicketRepository;
import com.pbshop.java.spring.maven.jpa.postgresql.user.User;
import com.pbshop.java.spring.maven.jpa.postgresql.user.UserRepository;

@ActiveProfiles("test")
@Transactional
class CommunityInquirySupportApiTest extends ApiIntegrationSupport {

    @Autowired
    private ObjectMapper objectMapper;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private BoardRepository boardRepository;

    @Autowired
    private CategoryRepository categoryRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private SellerRepository sellerRepository;

    @Autowired
    private PriceEntryRepository priceEntryRepository;

    @Autowired
    private ProductInquiryRepository productInquiryRepository;

    @Autowired
    private SupportTicketRepository supportTicketRepository;

    @Test
    void userCanCreateAndManageCommunityPostFlow() throws Exception {
        String userToken = createUserAndLogin("USER");
        Board board = createBoard("자유게시판", "free-board");

        MvcResult createPostResult = mockMvc.perform(post("/api/v1/boards/" + board.getId() + "/posts")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "title": "첫 게시글",
                                  "content": "커뮤니티 API 테스트"
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.boardId").value(board.getId()))
                .andReturn();

        long postId = objectMapper.readTree(createPostResult.getResponse().getContentAsString())
                .path("data")
                .path("id")
                .asLong();

        mockMvc.perform(get("/api/v1/boards/" + board.getId() + "/posts"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.items[0].id").value(postId));

        mockMvc.perform(post("/api/v1/posts/" + postId + "/like")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.liked").value(true))
                .andExpect(jsonPath("$.data.likeCount").value(1));

        MvcResult commentResult = mockMvc.perform(post("/api/v1/posts/" + postId + "/comments")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "content": "첫 댓글"
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.postId").value(postId))
                .andReturn();

        long commentId = objectMapper.readTree(commentResult.getResponse().getContentAsString())
                .path("data")
                .path("id")
                .asLong();

        mockMvc.perform(get("/api/v1/posts/" + postId))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.id").value(postId))
                .andExpect(jsonPath("$.data.comments[0].id").value(commentId));

        mockMvc.perform(delete("/api/v1/comments/" + commentId)
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.message").value("댓글이 삭제되었습니다."));

        mockMvc.perform(delete("/api/v1/posts/" + postId)
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.message").value("게시글이 삭제되었습니다."));
    }

    @Test
    void inquiryFlowSupportsSecretMaskingAndSellerAnswer() throws Exception {
        String userToken = createUserAndLogin("USER");
        String sellerToken = createUserAndLogin("SELLER");
        Category category = createCategory("노트북", "laptop");
        Product product = createProduct(category, "PB 노트북", "pb-laptop");
        Seller seller = createSeller("PB셀러", "pb-seller");
        createPriceEntry(product, seller, new BigDecimal("1299000"));

        MvcResult inquiryResult = mockMvc.perform(post("/api/v1/products/" + product.getId() + "/inquiries")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "title": "재고 문의",
                                  "content": "오늘 출고 가능한가요?",
                                  "isSecret": true
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.title").value("재고 문의"))
                .andReturn();

        long inquiryId = objectMapper.readTree(inquiryResult.getResponse().getContentAsString())
                .path("data")
                .path("id")
                .asLong();

        mockMvc.perform(get("/api/v1/products/" + product.getId() + "/inquiries"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data[0].content").value("비밀 문의입니다."));

        mockMvc.perform(post("/api/v1/inquiries/" + inquiryId + "/answer")
                        .header("Authorization", "Bearer " + sellerToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "answer": "오늘 바로 출고 가능합니다."
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.status").value("ANSWERED"))
                .andExpect(jsonPath("$.data.answer").value("오늘 바로 출고 가능합니다."));

        mockMvc.perform(get("/api/v1/inquiries/me")
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data[0].id").value(inquiryId))
                .andExpect(jsonPath("$.data[0].content").value("오늘 출고 가능한가요?"));

        mockMvc.perform(delete("/api/v1/inquiries/" + inquiryId)
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.message").value("문의가 삭제되었습니다."));
    }

    @Test
    void supportFlowSupportsUserTicketAndAdminManagement() throws Exception {
        String userToken = createUserAndLogin("USER");
        String adminToken = createUserAndLogin("ADMIN");

        MvcResult ticketResult = mockMvc.perform(post("/api/v1/support/tickets")
                        .header("Authorization", "Bearer " + userToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "category": "배송",
                                  "title": "배송 상태 확인",
                                  "content": "주문 건 배송 상태를 알고 싶습니다."
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.status").value("OPEN"))
                .andReturn();

        long ticketId = objectMapper.readTree(ticketResult.getResponse().getContentAsString())
                .path("data")
                .path("id")
                .asLong();

        mockMvc.perform(post("/api/v1/support/tickets/" + ticketId + "/reply")
                        .header("Authorization", "Bearer " + adminToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "content": "배송팀에 확인 후 안내드리겠습니다."
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.adminReply").value(true));

        mockMvc.perform(get("/api/v1/support/tickets/" + ticketId)
                        .header("Authorization", "Bearer " + userToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.id").value(ticketId))
                .andExpect(jsonPath("$.data.replies[0].adminReply").value(true));

        mockMvc.perform(get("/api/v1/admin/support/tickets")
                        .header("Authorization", "Bearer " + adminToken))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.items[0].id").value(ticketId));

        mockMvc.perform(patch("/api/v1/admin/support/tickets/" + ticketId + "/status")
                        .header("Authorization", "Bearer " + adminToken)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "status": "RESOLVED"
                                }
                                """))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.data.status").value("RESOLVED"));
    }

    private Board createBoard(String name, String slug) {
        Board board = new Board();
        board.setName(name);
        board.setSlug(slug);
        board.setDescription(name + " 설명");
        board.setActive(true);
        return boardRepository.save(board);
    }

    private Category createCategory(String name, String slug) {
        Category category = new Category();
        category.setName(name);
        category.setSlug(slug);
        category.setDepth(0);
        category.setSortOrder(0);
        category.setVisible(true);
        return categoryRepository.save(category);
    }

    private Product createProduct(Category category, String name, String slug) {
        Product product = new Product();
        product.setCategory(category);
        product.setName(name);
        product.setSlug(slug);
        product.setBrand("PB");
        product.setDescription("커뮤니티 테스트 상품");
        product.setThumbnailUrl("https://cdn.pbshop.local/" + slug + ".png");
        product.setReviewCount(0);
        product.setRatingAvg(BigDecimal.ZERO);
        product.setStatus("ACTIVE");
        return productRepository.save(product);
    }

    private Seller createSeller(String name, String code) {
        Seller seller = new Seller();
        seller.setName(name);
        seller.setCode(code);
        seller.setHomepageUrl("https://seller.pbshop.local/" + code);
        seller.setStatus("ACTIVE");
        return sellerRepository.save(seller);
    }

    private void createPriceEntry(Product product, Seller seller, BigDecimal price) {
        PriceEntry entry = new PriceEntry();
        entry.setProduct(product);
        entry.setSeller(seller);
        entry.setPrice(price);
        entry.setOriginalPrice(price);
        entry.setShippingFee(BigDecimal.ZERO);
        entry.setStockStatus("IN_STOCK");
        entry.setPurchaseUrl("https://seller.pbshop.local/p/" + product.getSlug());
        product.getPriceEntries().add(entry);
        priceEntryRepository.save(entry);
    }

    private String createUserAndLogin(String role) throws Exception {
        String email = role.toLowerCase() + "-" + UUID.randomUUID() + "@pbshop.test";
        User user = new User();
        user.setEmail(email);
        user.setPasswordHash(passwordEncoder.encode("Password123!"));
        user.setName(role.toLowerCase() + "-user");
        user.setPhone("01055556666");
        user.setRole(role);
        user.setStatus("ACTIVE");
        user.setEmailVerified(true);
        userRepository.save(user);

        MvcResult loginResult = mockMvc.perform(post("/api/v1/auth/login")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("""
                                {
                                  "email": "%s",
                                  "password": "Password123!"
                                }
                                """.formatted(email)))
                .andExpect(status().isOk())
                .andReturn();

        return objectMapper.readTree(loginResult.getResponse().getContentAsString())
                .path("data")
                .path("accessToken")
                .asText();
    }
}
